Rust backend

lucifer1004 opened the pull request:
Almost all features of the Julia backend have been ported. Many tests have been added.

This PR is ready for review.

certik commented:
Thanks!

I think we need to figure out an answer to this:

https://github.com/lfortran/lfortran/issues/816

in order to decide how to proceed here.

lucifer1004 commented:
As I say in #816, a solid framework is needed to switch to the plugin route. So I think we can merge the Rust backend into LFortran for now, until there is another person who starts a new backend. Personally, I am not going to add more backends at the moment.

certik commented:
It takes 5h for a single person to write the initial backend but it adds days of extra work for the whole team to maintain it. Currently, we are focusing on getting LFortran from alpha to beta, so I don't want us to spread too thin and delay that.

We need to figure out how to maintain such backends and tests separately, so that we can move quickly with LFortran, and then people interested in these extra backends can update them without slowing the main development.

czgdp1807 commented:
Correct. I think everything depends on the stability of `libasr`. If it will be more stable then it will be very easy to scale the number of backends. Right now its updating really fast. However more backends helps us in designing ASR better because we have to satisfy more constraints (`GoTo` update due to C backend is a good example for this).

lucifer1004 commented:
I implemented this backend since

- There is issue #254, which at least showed that at some point the Rust backend was taken into serious consideration
- Rust is also a language that I use often and love.

The current implementation has passed the tests I added (by actually compiling), except for one or two I explicitly marked. Also, I think Rust, as a language gaining more and more popularity and having some unique features, can help us to improve the ASR design. For example, the idea of lifetime, ownership, and traits are all worth considering.

czgdp1807 commented:
Well we have just to consider the pros and cons of adding at the current stage of development of LCompilers. I am not strongly opinionated on either side. Just properly considering all things will help us make a better decision. Some of the points to consider,

1. Maintenance costs  - Is it easy enough to maintain the new backend?
2. What advantages does the new backend provides as compared to the already available ones?
3. How stable is the `libasr` when new backend is added?
4. How much the new backend slows down the CI tests?
5. How much impact can the new backend create if added? Will it give a lot more exposure to LCompilers.

The above questions are not in any specific priority order. Just weighing the answers to each of these will help in making a better decision.

certik commented:
@lucifer1004 the Rust backend is definitely under serious consideration. The question is not if (the answer is of course yes!), the question is how. See https://github.com/lfortran/lfortran/issues/816#issuecomment-1263758457 for more details.

certik commented:
We have to maintain this in a separate repository, so that it doesn't slow us down when delivering beta of LFortran, which is the absolute highest priority now.
